### PR TITLE
support for 8th creature (Factory)

### PR DESCRIPTION
**Pull request by Laserlicht**
Allows to add optional 8th dwelling to town.

Not sure if there are still any unknown side effects... But seems to work.

Can anyone test it? Download HotA beta branch mod.

Change in town.json:
```
			"creatures" : [
				[ "halfling", "halflingGrenadier" ],
				[ "hotaMechanic", "hotaEngineer" ],
				[ "armadillo", "armadilloUp" ],
				[ "automaton", "automatonUp" ],
				[ "sandworm", "olgoiKhorkhoi" ],
				[ "hotaGunman", "hotaGunmanUp" ],
				[ "hotaCouatl", "hotaCoutalUp" ],
				[ "hotaDread", "hotaJugger" ],
				
			],
```

replace all `dwellingLvl7-2` to `dwellingLvl8` and all `dwellingUpLvl-7` to `dwellingUpLvl8`.

grafik
grafik

### CODE REVIEW

**Comment by DjWarmonger** [comment]
Uhhh, dwellings in fort view are mixed up. Gantry is listed twice, while it should be for Dreadnought only.

**Comment by Laserlicht** [comment]
> Uhhh, dwellings in fort view are mixed up. Gantry is listed twice, while it should be for Dreadnought only.



> Uhhh, dwellings in fort view are mixed up. Gantry is listed twice, while it should be for Dreadnought only.

Yes. Should be corrected in Mod.

**Comment by Laserlicht** [comment]
Seems that there is still an conflict with ToW stuff. Do anyone know how it works?

**Comment by misiokles** [comment]
Every town dwelling is connected with id. So, for example Portal of Glory had id: 43, but id: 44 is connected with second upgrade of lvl1 and id: 50 is connected with second upgrade of lvl7. So, id: 51 is third upgrade of lvl2 and round and round. Check third-upgrades json
https://discord.com/channels/298106089885401090/298108620317655052/1188403725740556329

So 8th lvl dwelling won't have free id.

**Comment by Laserlicht** [comment]
Seems to work now.

No errors. Factory and other towns seems to work as expected. Altar of Sacrifice in Dungeon also works.

Can someone other test it?

grafik

**Comment by ToRRent1812** [comment]
I know it's a little off topic, but does this will also show alternative creatures to recruit?

File: lib/constants/EntityIdentifiers.h [hunk 1]
```diff
@@ -313,6 +314,44 @@ class BuildingIDBase : public IdentifierBase
...
+		std::vector<Type> dwellings = { DWELL_LVL_1, DWELL_LVL_2, DWELL_LVL_3, DWELL_LVL_4, DWELL_LVL_5, DWELL_LVL_6, DWELL_LVL_7, DWELL_LVL_8 };
+		std::vector<Type> dwellingsUp = { DWELL_LVL_1_UP, DWELL_LVL_2_UP, DWELL_LVL_3_UP, DWELL_LVL_4_UP, DWELL_LVL_5_UP, DWELL_LVL_6_UP, DWELL_LVL_7_UP, DWELL_LVL_8_UP };
+		return {dwellings, dwellingsUp};
+	}
+
+public:
+	static Type getDwellingFromLevel(int level, bool up = false)
+	{
+		return getDwellings()[up ? 1 : 0][level];
+	}
```
[the remark this answers is not in the document]
Comment (Laserlicht) [reply lib/constants/EntityIdentifiers.h]: Yes. Changed it... Thanks for explaining. :)

File: lib/constants/NumericConstants.h [hunk 2]
```diff
@@ -22,7 +22,7 @@ namespace GameConstants
 
 	constexpr int ALL_PLAYERS = 255; //bitfield
 
-	constexpr int CREATURES_PER_TOWN = 7; //without upgrades
+	constexpr int CREATURES_PER_TOWN = 8; //without upgrades
```
[the remark this answers is not in the document]
Comment (Laserlicht) [reply lib/constants/NumericConstants.h:25]: Still works fine. Tested initial fill and refill (with same/new creature) slot at new week.

Yes: If someone want to do a mod that places 8th creature in standard towns he also have to remove portal of summoning to do this on dungeon.

**Comment by Laserlicht** [comment]
> I know it's a little off topic, but does this will also show alternative creatures to recruit?

What exactly do you mean? Alternative creatures (e.g. pirates, tow mod, ...) are still working. This PR only add the possibility for 8th dwelling and creature slot.

**Comment by ToRRent1812** [comment]
> > I know it's a little off topic, but does this will also show alternative creatures to recruit?
> 
> What exactly do you mean? Alternative creatures (e.g. pirates, tow mod, ...) are still working. This PR only add the possibility for 8th dwelling and creature slot.

Oh, I thought, it also adds new unit slot in fort

**Comment by Laserlicht** [comment]
Additional info:

For 8th slot the id's 150 and 151 are used for the sprite image of 8th dwelling in town-hall. That is to avoid conflicts with existing mods (e.g. ToW) using smaller numbers.

To avoid blowing up def file: It should be in json format, with only the necessary sprites, e.g.:
```
{
	"basepath" : "factory/",
	"images" :
	[
		{ "frame" :  0, "file" : "hallfactA000.bmp"},
		{ "frame" :  1, "file" : "hallfactA001.bmp"},
		{ "frame" :  2, "file" : "hallfactA002.bmp"},
		{ "frame" :  3, "file" : "hallfactA003.bmp"},
		{ "frame" :  4, "file" : "hallfactA004.bmp"},
		{ "frame" :  5, "file" : "hallfactA005.bmp"},
		{ "frame" :  6, "file" : "hallfactA006.bmp"},
		{ "frame" :  7, "file" : "hallfactA007.bmp"},
		{ "frame" :  8, "file" : "hallfactA008.bmp"},
		{ "frame" :  9, "file" : "hallfactA009.bmp"},
		{ "frame" :  10, "file" : "hallfactA010.bmp"},
		{ "frame" :  11, "file" : "hallfactA011.bmp"},
		{ "frame" :  12, "file" : "hallfactA012.bmp"},
		{ "frame" :  13, "file" : "hallfactA013.bmp"},
		{ "frame" :  14, "file" : "hallfactA014.bmp"},
		{ "frame" :  15, "file" : "hallfactA015.bmp"},
		{ "frame" :  16, "file" : "hallfactA016.bmp"},
		{ "frame" :  17, "file" : "hallfactA021.bmp"},
		{ "frame" :  18, "file" : "hallfactA018.bmp"},
		{ "frame" :  19, "file" : "hallfactA019.bmp"},
		{ "frame" :  20, "file" : "hallfactA020.bmp"},
		{ "frame" :  21, "file" : "hallfactA006.bmp"},
		{ "frame" :  22, "file" : "hallfactA017.bmp"},
		{ "frame" :  26, "file" : "hallfactA026.bmp"},
		{ "frame" :  30, "file" : "hallfactA030.bmp"},
		{ "frame" :  31, "file" : "hallfactA031.bmp"},
		{ "frame" :  32, "file" : "hallfactA032.bmp"},
		{ "frame" :  33, "file" : "hallfactA033.bmp"},
		{ "frame" :  34, "file" : "hallfactA034.bmp"},
		{ "frame" :  35, "file" : "hallfactA035.bmp"},
		{ "frame" :  36, "file" : "hallfactA036.bmp"},
		{ "frame" :  37, "file" : "hallfactA037.bmp"},
		{ "frame" :  38, "file" : "hallfactA038.bmp"},
		{ "frame" :  39, "file" : "hallfactA039.bmp"},
		{ "frame" :  40, "file" : "hallfactA040.bmp"},
		{ "frame" :  41, "file" : "hallfactA041.bmp"},
		{ "frame" :  42, "file" : "hallfactA042.bmp"},
		{ "frame" :  43, "file" : "hallfactA043.bmp"},
		{ "frame" :  150, "file" : "hallfactA022.bmp"},
		{ "frame" :  151, "file" : "hallfactA023.bmp"}
	]
}
```

File: lib/constants/EntityIdentifiers.h [hunk 3]
```diff
@@ -313,6 +314,44 @@ class BuildingIDBase : public IdentifierBase
...
+
+	static int getLevelFromDwelling(BuildingIDBase dwelling)
+	{
+		for(int i = 0; i < 2; i++)
```
[the remark this answers is not in the document]
Comment (Laserlicht) [reply lib/constants/EntityIdentifiers.h:333]: Yes tested it again. Works as expected.

**Comment by krs0** [comment]
So could it show 2 lvl7 HotA? Or one must be defined as lvl8?

**Comment by IvanSavenko** [comment]
> So could it show 2 lvl7 HotA? Or one must be defined as lvl8?

These 'levels' are not necessarily creature levels.
More like dwelling slots with independent growth. (as opposed to 'upgrades' - slots with shared growth)

External Golem Factory and Elemental Conflux use 4 such "levels", even though creature levels are completely different. Portal of Summoning in Dungeon was always using this "8th level", even though it could summon anything.

File: client/windows/CCastleInterface.cpp [hunk 4]
```diff
@@ -1891,7 +1891,7 @@ const CBuilding * CFortScreen::RecruitArea::getMyBuilding()
 	{
 		if (town->hasBuilt(myID))
 			build = town->town->buildings.at(myID);
-		myID.advance(GameConstants::CREATURES_PER_TOWN);
+		myID.advance(town->town->creatures.size());
```
Comment (IvanSavenko) [inline client/windows/CCastleInterface.cpp:1894]: I think this would break upgraded buildings in fort?
Can you check that Fort window works correctly with upgraded buildings, both for tiers 1-7 and for tier 8?
Because for tiers 1-7 advancing needs to be done by 7 (ID of upgraded dwelling - ID of downgrade) and for new tier 8 - only by 1.

I think this needs your new `getDwellingFromLevel` method?

Comment (Laserlicht) [reply client/windows/CCastleInterface.cpp:1894]: Fort is absolutely fine for my. With 7 and 8 creatures.
Tested in Fort using vcmibuild (also tested with "empty" town):
- Buying (and selecting other level)
- Right click
